File: dexfile/src/test/java/com/github/netomi/bat/dexfile/EncodedFieldTest.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile;

import com.github.netomi.bat.dexfile.io.DexDataInput;
import com.github.netomi.bat.dexfile.io.DexDataOutput;
import org.junit.jupiter.api.Test;

import java.util.EnumSet;
import java.util.function.Consumer;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.*;

public class EncodedFieldTest
extends      DexContentTest<EncodedField>
{
    @Override
    public EncodedField[] getTestInstances() {
        return new EncodedField[] {
            EncodedField.of(1, Visibility.PUBLIC, FieldModifier.FINAL),
            EncodedField.of(2, Visibility.PRIVATE),
            EncodedField.of(65535, Visibility.PACKAGE_PRIVATE, FieldModifier.VOLATILE, FieldModifier.SYNTHETIC),
        };
    }

    @Override
    public Function<DexDataInput, EncodedField> getFactoryMethod() {
        return (input) -> EncodedField.readContent(input, 0);
    }

    @Override
    public Consumer<DexDataOutput> getWriteMethod(EncodedField data) {
        return (output) -> data.write(output, 0);
    }

    @Test
    public void inputChecking() {
        assertThrows(IllegalArgumentException.class, () -> {
            EncodedField.of(-1, Visibility.PUBLIC, FieldModifier.FINAL);
        });
    }

    @Test
    public void getter() {
        EncodedField[] data = getTestInstances();

        assertEquals(1, data[0].getFieldIndex());
        assertEquals(Visibility.PUBLIC, data[0].getVisibility());
        assertEquals(EnumSet.of(FieldModifier.FINAL), data[0].getModifiers());
    }

    @Test
    public void equals() {
        EncodedField e1 = EncodedField.of(1, Visibility.PUBLIC, FieldModifier.FINAL);
        EncodedField e2 = EncodedField.of(2, Visibility.PRIVATE, FieldModifier.STATIC);
        EncodedField e3 = EncodedField.of(1, Visibility.PUBLIC, FieldModifier.FINAL);

        assertEquals(e1, e1);
        assertNotEquals(e1, e2);
        assertEquals(e1, e3);
    }
}

File: dexfile/src/test/kotlin/com/github/netomi/bat/dexfile/EncodedFieldTest.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile

import com.github.netomi.bat.dexfile.EncodedField.Companion.of
import com.github.netomi.bat.dexfile.EncodedField.Companion.readContent
import com.github.netomi.bat.dexfile.io.DexDataInput
import com.github.netomi.bat.dexfile.io.DexDataOutput
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.util.*
import java.util.function.Consumer
import java.util.function.Function

class EncodedFieldTest : DexContentTest<EncodedField>() {
    override val testInstances: Array<EncodedField>
        get() = arrayOf(
                    of(1, Visibility.PUBLIC, FieldModifier.FINAL),
                    of(2, Visibility.PRIVATE),
                    of(65535, Visibility.PACKAGE_PRIVATE, FieldModifier.VOLATILE, FieldModifier.SYNTHETIC)
                )

    override val factoryMethod: Function<DexDataInput, EncodedField>
        get() = Function { input -> readContent(input, 0) }

    override fun getWriteMethod(data: EncodedField): Consumer<DexDataOutput> {
        return Consumer { output -> data.write(output, 0) }
    }

    @Test
    fun inputChecking() {
        assertThrows(IllegalArgumentException::class.java) { of(-1, Visibility.PUBLIC, FieldModifier.FINAL) }
    }

    @Test
    fun getter() {
        val data = testInstances
        assertEquals(1, data[0].fieldIndex)
        assertEquals(Visibility.PUBLIC, data[0].visibility)
        assertEquals(EnumSet.of(FieldModifier.FINAL), data[0].modifiers)
    }

    @Test
    fun equals() {
        val e1 = of(1, Visibility.PUBLIC, FieldModifier.FINAL)
        val e2 = of(2, Visibility.PRIVATE, FieldModifier.STATIC)
        val e3 = of(1, Visibility.PUBLIC, FieldModifier.FINAL)
        assertEquals(e1, e1)
        assertNotEquals(e1, e2)
        assertEquals(e1, e3)
    }
}